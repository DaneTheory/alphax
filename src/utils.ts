/**
 * data types check util
 * Modified from https://github.com/sindresorhus/is/blob/master/index.js
 */
import * as File from 'vinyl'
import { RenameConfig } from './alphax'

const toString = Object.prototype.toString
const getObjectType = x => toString.call(x).slice(8, -1)
const isOfType = type => x => typeof x === type // eslint-disable-line valid-typeof
const isObjectOfType = type => x => getObjectType(x) === type

export const isBoolean = x => x === true || x === false

export const isSymbol = isOfType('symbol')

export const isInteger = Number.isInteger

export const isSafeInteger = Number.isSafeInteger

export const isString = isOfType('string')

export const isUndefined = isOfType('undefined')

export const isNull = x => x === null

export const isNullOrUndefined = x => isUndefined(x) || isNull(x)

export const isFunction = isOfType('function')

export const isBuffer = Buffer.isBuffer

export const isGeneralizedObject = x => typeof x === 'object'

export const isPlainObject = isObjectOfType('Object')

export const isObject = x => !isNullOrUndefined(x) && (isFunction(x) || isGeneralizedObject(x))

const isValidLength = x => isSafeInteger(x) && x > -1;

export const isArray = Array.isArray

export const isArrayLike = x => !isNullOrUndefined(x) && !isFunction(x) && isValidLength(x.length);

export const isNativePromise = isObjectOfType('Promise')

const hasPromiseAPI = x => isObject(x) && isFunction(x.then) && isFunction(x.catch)

export const isPromise = x => isNativePromise(x) || hasPromiseAPI(x)

export const isIterable = x => !isNullOrUndefined(x) && isFunction(x[Symbol.iterator])

export const isGenerator = x => isIterable(x) && isFunction(x.next) && isFunction(x.throw)

export const isRegExp = isObjectOfType('RegExp')

export function curryFileTransformer(fn) {
  return function (file: File, cb) {
    const res = fn(file)
    if (isPromise(res)) {
      res.then(asyncRes => {
        if (asyncRes === null) {
          cb()
        } else {
          cb(null, file)
        }
      }).catch(error => {
        console.log(error)
      })
    } else {
      if (res === null) {
        cb()
      } else {
        cb(null, file)
      }
    }
  }
}

export function getRenamerByConfig(renameConfig: RenameConfig, context: any): (filepath: string) => string {
  return function (filepath: string) {
    Object.keys(renameConfig).forEach(pattern => {
      let condition = renameConfig[pattern]
      if (/(\?|\=)/.test(condition)) {
        try {
          condition = evaluate(condition, context)
        } catch (error) {
          console.log(error)
        }
      }
      if (isNull(condition)) {
        return
      }
      filepath = filepath.replace(new RegExp(pattern, 'g'), condition)
    })
    return filepath
  }
}

export function evaluate(exp: string | boolean, data: any) {
  /* eslint-disable no-new-func */
  const fn = new Function('data', 'with (data) { return ' + exp + '}')
  try {
    return fn(data)
  } catch (e) {
    throw new Error('Error when evaluating filter condition: ' + exp)
  }
}
