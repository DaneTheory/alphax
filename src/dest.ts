import * as vinyl from 'vinyl-fs'
import * as File from 'vinyl'
import * as es from 'event-stream'
import { isString, isFunction, isArray, isUndefined, isPromise } from './utils'
import ReadWriteStream = NodeJS.ReadWriteStream;

type Glob = string | string[]

interface Options extends vinyl.SrcOptions {
  transformer?: (file: File) => void
}

export default async function dest(source: Glob,
                                   target: string,
                                   options?: Options = {}): Promise<ReadWriteStream> {

  console.log(source)

  if (!isString(source) && !isArray(source)) {
    throw new Error('Expected "source" to be string or array.')
  }
  if (!isString(target)) {
    throw new Error('Expected "target" to be string.')
  }

  const globs = source
  const stream: ReadWriteStream = vinyl.src(globs, options)
  const { transformer } = options

  if (!isUndefined(transformer) && !isFunction(transformer)) {
    stream.emit('unexpected-transformer')
  }

  const transform = (file: File, cb) => {
    let transformRes
    if (transformer) {
      try {
        transformRes = transformer(file)
      } catch (error) {
        stream.emit('transform-error', error)
      }
    }

    if (isPromise(transformRes)) {
      return transformRes.then(() => {
        cb(null, file)
      })
    }

    cb(null, file)
  }

  stream
    .pipe(es.map(transform))
    .pipe(vinyl.dest(target))

  await new Promise((resolve, reject) => {
    stream.on('end', () => resolve(stream))
    stream.on('error', reject)
  })
}
