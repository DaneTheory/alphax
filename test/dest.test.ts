import * as path from 'path'
import * as File from 'vinyl'
import dest from '../src/dest'
import { SRC_DIR, getDistDir, globDir } from './utils'

describe('dest', () => {

  test('basic', async () => {
    const DIST_DIR = getDistDir('basic')
    await dest(SRC_DIR + '/**', DIST_DIR)
    const files = await globDir(DIST_DIR, { baseDir: DIST_DIR })
    expect(files).toMatchSnapshot()
  })

  test('ignore', async () => {
    const DIST_DIR = getDistDir('ignore')
    await dest(
      [
        SRC_DIR + '/**',
        '!' + SRC_DIR + '/d.js',
        '!' + SRC_DIR + '/a/b.js'
      ],
      DIST_DIR,
      {
        allowEmpty: true
      }
    )
    const files = await globDir(DIST_DIR, { baseDir: DIST_DIR })
    expect(files).toMatchSnapshot()
  })

  test('rename', async () => {
    const DIST_DIR = getDistDir('rename')
    await dest(
      [
        SRC_DIR + '/**'
      ],
      DIST_DIR,
      {
        allowEmpty: true,
        transformer: (file: File) => {
          // Rename all files to ts
          file.path = path.join(file.base, (<string>file.relative).replace('.js', '.ts'))
          // Rename dir
          file.path = path.join(file.base, (<string>file.relative).replace('a', 'A'))
        }
      }
    )
    const files = await globDir(DIST_DIR, { baseDir: DIST_DIR })
    expect(files).toMatchSnapshot()
  })

})
