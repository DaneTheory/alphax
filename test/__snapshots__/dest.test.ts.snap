// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dest basic 1`] = `
Array [
  "a",
  "a/b",
  "a/b.js",
  "a/b/c.js",
  "d.js",
]
`;

exports[`dest ignore 1`] = `
Array [
  "a",
  "a/b",
  "a/b/c.js",
]
`;

exports[`dest rename 1`] = `
Array [
  "A",
  "A/b",
  "A/b.ts",
  "A/b/c.ts",
  "d.ts",
]
`;
